graylog2-server/src/main/java/org/graylog2/contentpacks/ContentPackPersistenceService.java

```diff
@@ -0,0 +1,117 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.contentpacks;
+
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.mongodb.BasicDBObject;
+import org.bson.types.ObjectId;
+import org.graylog2.bindings.providers.MongoJackObjectMapperProvider;
+import org.graylog2.contentpacks.model.ContentPack;
+import org.graylog2.contentpacks.model.Identified;
+import org.graylog2.contentpacks.model.ModelId;
+import org.graylog2.contentpacks.model.Revisioned;
+import org.graylog2.database.MongoConnection;
+import org.mongojack.DBCursor;
+import org.mongojack.DBQuery;
+import org.mongojack.JacksonDBCollection;
+import org.mongojack.WriteResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.Set;
+
+@Singleton
+public class ContentPackPersistenceService {
+    public static final String COLLECTION_NAME = "content_packs";
+
+    private final JacksonDBCollection<ContentPack, ObjectId> dbCollection;
+    private static final Logger LOG = LoggerFactory.getLogger(ContentPackPersistenceService.class);
+
+    @Inject
+    public ContentPackPersistenceService(final MongoJackObjectMapperProvider mapperProvider,
+                                         final MongoConnection mongoConnection) {
+        this(JacksonDBCollection.wrap(mongoConnection.getDatabase().getCollection(COLLECTION_NAME),
+                ContentPack.class, ObjectId.class, mapperProvider.get()));
+    }
+
+    ContentPackPersistenceService(final JacksonDBCollection<ContentPack, ObjectId> dbCollection) {
+        this.dbCollection = dbCollection;
+
+        dbCollection.createIndex(new BasicDBObject(Identified.FIELD_META_ID, 1).append(Revisioned.FIELD_META_REVISION, 1), new BasicDBObject("unique", true));
```

Comment:
The server still doesn't start with legacy content packs in the database due to a `DuplicateKeyError`. This happens when this code is running before the migration that fixes up the legacy content packs has run.

We could add something like the following to ignore errors during index creation. This allows the migration to run and the index will be created during a next server restart.

```diff
diff --git graylog2-server/src/main/java/org/graylog2/contentpacks/ContentPackPersistenceService.java graylog2-server/src/main/java/org/graylog2/contentpacks/ContentPackPersistenceService.java
index 3e35b8411..beb22fd13 100644
--- graylog2-server/src/main/java/org/graylog2/contentpacks/ContentPackPersistenceService.java
+++ graylog2-server/src/main/java/org/graylog2/contentpacks/ContentPackPersistenceService.java
@@ -20,6 +20,7 @@ import com.google.common.collect.ImmutableCollection;
 import com.google.common.collect.ImmutableMultimap;
 import com.google.common.collect.ImmutableSet;
 import com.mongodb.BasicDBObject;
+import com.mongodb.DuplicateKeyException;
 import org.bson.types.ObjectId;
 import org.graylog2.bindings.providers.MongoJackObjectMapperProvider;
 import org.graylog2.contentpacks.model.ContentPack;
@@ -58,7 +59,11 @@ public class ContentPackPersistenceService {
     ContentPackPersistenceService(final JacksonDBCollection<ContentPack, ObjectId> dbCollection) {
         this.dbCollection = dbCollection;
 
-        dbCollection.createIndex(new BasicDBObject(Identified.FIELD_META_ID, 1).append(Revisioned.FIELD_META_REVISION, 1), new BasicDBObject("unique", true));
+        try {
+            dbCollection.createIndex(new BasicDBObject(Identified.FIELD_META_ID, 1).append(Revisioned.FIELD_META_REVISION, 1), new BasicDBObject("unique", true));
+        } catch (DuplicateKeyException e) {
+            // Ignore - this can happen if this runs before the migration of old content packs
+        }
     }
 
     public Set<ContentPack> loadAll() {
```

This makes the server start for me.